include/swift/SIL/MemoryLifetime.h

```diff
@@ -0,0 +1,374 @@
+                                      SmallVectorImpl<SILValue> &collectedVals,
+                                      SubLocationMap &subLocationMap);
+
+  /// Helper function called by analyzeLocation to create a sub-location for
+  /// and address projection and check all of its uses.
+  bool analyzeAddrProjection(
+      SILValue projection, unsigned parentLocIdx,unsigned fieldNr,
+      SmallVectorImpl<SILValue> &collectedVals, SubLocationMap &subLocationMap);
+
+  /// Calculates Location::numFieldsNotCoveredBySubfields
+  void initFieldsCounter(Location &loc);
+
+  /// Handles locations of the predecessor's terminator, which are only valid
+  /// in \p block.
+  /// Example: @out results of try_apply. They are only valid in the
+  /// normal-block, but not in the throw-block.
+  void setBitsOfPredecessor(Bits &bits, SILBasicBlock *block);
+
+  /// Sets up the BlockState datastructures and associates all basic blocks with
+  /// a state.
+  void initBlockStates();
+
+  /// Initializes the data flow bits sets in the block states for all blocks.
+  void initDataflow();
+
+  /// Initializes the data flow bits sets in the block state for a single block.
+  void initDataflowInBlock(BlockState &state);
+
+  /// Helper function to set bits for function arguments and returns.
+  void setFuncOperandBits(BlockState &state, Operand &op,
+                          SILArgumentConvention convention,
+                          bool isTryApply);
+
+  /// Calculates the BlockState::reachableFromEntry flags.
+  void entryReachabilityAnalysis();
+
+  /// Calculates the BlockState::exitReachable flags.
+  void exitReachableAnalysis();
+
+  /// Derives the block exit sets from the entry sets by applying the gen and
+  /// kill sets.
+  void solveDataflowForward();
+
+  /// Derives the block entry sets from the exit sets by applying the gen and
+  /// kill sets.
+  void solveDataflowBackward();
+
+  /// Do a block-local processing for all locations in singleBlockLocations.
+  ///
+  /// First, initializes all locations which are alive in a block and then
+  /// calls \p handlerFunc for the block.
+  void handleSingleBlockLocations(
+                       std::function<void (SILBasicBlock *block)> handlerFunc);
+
+  MemoryLifetime(SILFunction *F) : F(F) { }
+
+public:
+
+  /// Debug dump the MemoryLifetime internals.
+  void dump() const;
+
+  /// Debug dump a bit set .
+  static void dumpBits(const Bits &bits);
+
+};
+
+/// A utility for verifying memory lifetime.
+///
+/// The MemoryLifetime utility checks the lifetime of memory locations.
+/// This is limited to memory locations which are guaranteed to be not aliased,
+/// like @in or @inout parameters. Also, alloc_stack locations are handled.
+///
+/// In addition to verification, the MemoryLifetime class can be used as utility
+/// (e.g. base class) for optimizations, which need to compute memory lifetime.
+class MemoryLifetimeVerifier : public MemoryLifetime {
```

Comment:
Does this have to be a subclass? Can we use composition instead?

Reply:
I also meant to comment on this. Making this a subclass is deeply confusing to me. Please don't use inheritance as a mechanism for code reuse! In fact, it will be incredibly annoying that everything in the MemoryLifetime class is under protected visibility.

Reply:
Let me clarify the comment on "not using inheritance". In the best design, each class stands on its own, so it's easy to adapt and compose for new situations.

First, basic principles:

A class definition aggregates state that has the same identity and lifetime. Any operation that is meaningful to use outside that class should be provided by a public interface. Operations that can be expressed solely via that interface should be usually defined outside the class.

Inheritance is useful for polymorphism (e.g. passing any kind of instruction to a routine that doesn't care about the opcode), but beyond that it's usually anti-pattern, in that it makes the code harder to understand, compose, and debug.

Here is a template that works for designing an analysis utility:
- Separate the analysis result from the transient state
- Define a short-lived analysis class manages transient state and produces the result
- Use an rvalue method ("Result && analyze() &&") that destroys the transient state when returning its result.

Example usage:
"Result result = Analysis(environment).analyze(inputs)"

e.g. see DominatedAccessAnalysis::analyze().

The way "protected" is being used in this PR is to distinguish between the operations that are valid *during* analysis vs. operations that are valid on the result of the analysis. To the extent that it is important, that should be done by separating the analysis interface from the result interface.

As a counter example, I have been attempting to debug EscapeAnalysis utility. This involves hacking in code that detects certain pattens. The fact that all the interfaces are declared protected makes it extremely cumbersome to work with for no benefit.

Another loosely related style point: I generally dislike nested class definitions. It's not too bad here because they are fairly small, but it usually does more harm than good for understanding and hacking on the code.

Reply:
A big problem with this approach of organizing the analysis logic into "protected" methods is that it's impossible to hack on the implementation without constantly updating the header. The header should only contain the class layout and public interface needed to access the result of the analysis. The analysis logic should be confined to the .cpp file as either "static" functions or internal classes that manage all the transient state. (This is another annoying thing about EscapeAnalysis).